hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py

```diff
@@ -0,0 +1,351 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from calendar import monthrange
+
+from dateutil.relativedelta import relativedelta
+from dateutil.rrule import (
+    DAILY,
+    FR,
+    MO,
+    MONTHLY,
+    SA,
+    SU,
+    TH,
+    TU,
+    WE,
+    WEEKLY,
+    YEARLY,
+    rrule,
+)
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+MONTHS = {
+    "january": 31,
+    "february": 28,
+    "march": 31,
+    "april": 30,
+    "may": 31,
+    "june": 30,
+    "july": 31,
+    "august": 31,
+    "september": 30,
+    "october": 31,
+    "november": 30,
+    "december": 31,
+}
+
+DAYS = {
+    "mon": MO,
+    "tue": TU,
+    "wed": WE,
+    "thu": TH,
+    "fri": FR,
+    "sat": SA,
+    "sun": SU,
+}
+
+WEEKS = {
+    "first": 1,
+    "second": 2,
+    "third": 3,
+    "last": 4,
+}
+
+
+class HRTimeSheetRecurrence(models.Model):
+    _name = "hr.timesheet.recurrence"
+    _description = "HR TimeSheet Recurrence"
+
+    task_ids = fields.One2many("hr.employee", "recurrence_id")
+    next_recurrence_date = fields.Date()
+    recurrence_left = fields.Integer(string="Number of tasks left to create")
+
+    repeat_interval = fields.Integer(string="Repeat Every", default=1)
+    repeat_unit = fields.Selection(
+        [
+            ("day", "Days"),
+            ("week", "Weeks"),
+            ("month", "Months"),
+            ("year", "Years"),
+        ],
+        default="week",
+    )
+    repeat_type = fields.Selection(
+        [
+            ("forever", "Forever"),
+            ("until", "End Date"),
+            ("after", "Number of Repetitions"),
+        ],
+        default="forever",
+        string="Until",
+    )
+    repeat_until = fields.Date(string="End Date")
+    repeat_number = fields.Integer(string="Repetitions")
+
+    repeat_on_month = fields.Selection(
+        [
+            ("date", "Date of the Month"),
+            ("day", "Day of the Month"),
+        ]
+    )
+
+    repeat_on_year = fields.Selection(
+        [
+            ("date", "Date of the Year"),
+            ("day", "Day of the Year"),
+        ]
+    )
+
+    mon = fields.Boolean(string="Mon")
+    tue = fields.Boolean(string="Tue")
+    wed = fields.Boolean(string="Wed")
+    thu = fields.Boolean(string="Thu")
+    fri = fields.Boolean(string="Fri")
+    sat = fields.Boolean(string="Sat")
+    sun = fields.Boolean(string="Sun")
+
+    repeat_day = fields.Selection([(str(i), str(i)) for i in range(1, 32)])
```

Comment:
Here too better use integer field with borders

Reply:
okay,I'll fix it